tests/plugins/test_pin_operator_digests.py

```diff
@@ -1206,3 +1223,148 @@ def test_pullspecs_replacements(self, tmpdir, docker_tasker):
+         False),
+        ({'pullspec_replacements': [],
+          'append': {'test': ['v1', 'v2']}},
+         True),
+        ({'pullspec_replacements': [],
+          'append': {'test': {'test2': ['v1', 'v2']}}},
+         True),
+        ({'pullspec_replacements': [],
+          'append': {'test': {'test2': 'must_be_a_list'}}},
+         False),
+        ({'pullspec_replacements': [],
+          'append': {'test': {'test2': [{'no_dict_here': ':-('}, 'v2']}}},
+         False),
+        ({'pullspec_replacements': [],
+          'append': {'test': {'test2': [['no_nested_list']]}}},
+         False),
+        ({'pullspec_replacements': [],
+          'update': []},
+         False),
+        ({'pullspec_replacements': [],
+          'update': {'test': 'val'}}, True),
+        ({'pullspec_replacements': [],
+          'update': {'test': {'test2': 'val'}}}, True),
+        ({'pullspec_replacements': [],
+          'update': {'test': ['list_not_allowed_in_recursive_update']}},
+         False),
+    ])
+    def test_operator_csv_modification_schema(self, data, valid):
+        """Unittests for operators CSV modification schema validating user input"""
+        schema = load_schema(
+            'atomic_reactor',
+            'schemas/operator_csv_modifications.json'
+        )
+        if valid:
+            validate_with_schema(data, schema)
+        else:
+            with pytest.raises(OsbsValidationException):
+                validate_with_schema(data, schema)
+
+    @responses.activate
+    def test_duplicated_pullspec_replacements(self, tmpdir, docker_tasker):
+        """Fail when duplicated pullspecs are detected"""
+        test_pullspec = "thesameimage:v1"
+        test_url = "https://example.com/modifications.json"
+        modification_data = {
+            "pullspec_replacements": [
+                {"original": test_pullspec, 'new': 'different:v1', 'pinned': False},
+                {"original": test_pullspec, 'new': 'different:v2', 'pinned': False},
+            ]
+        }
+        responses.add(responses.GET, test_url, json=modification_data)
+
+        self._test_assert_error(
+            tmpdir=tmpdir,
+            docker_tasker=docker_tasker,
+            test_url=test_url,
+            pull_specs=[test_pullspec],
+            exc_msg=f"Provided CSV modifications contain duplicated "
+                    f"original entries in pullspec_replacement: {test_pullspec}",
+        )
+
+    @pytest.mark.parametrize('mods,exc_msg', [
+        (
+            {'update': {'spec': {'test': 'something'}}},
+            (
+                "Operator CSV attributes [('spec', 'test')] are not allowed to be modified "
+                "by service configuration. Attributes allowed for modification "
+                "are {('spec', 'version')}"
+            )
+        ), (
+            {'append': {'spec': {'test': ['something']}}},
+            (
+                "Operator CSV attributes [('spec', 'test')] are not allowed to be modified "
```

Comment:
This error message is very ugly, I can replace `[('spec', 'test')]` with `"spec.test"`, eventually if there is a `.` in key, I can escape it `["metadata", "olm.skipRanges"]` -> `"metadata.olm\.skipRanges"`

Wdyt?

Reply:
ACK

Reply:
Changed